Address Safer CPP warnings in ImageCapture.cpp

eric-carlson opened the pull request:
#### ae7cc8174a2ee1e6a43cc7063883e9e0cdc44992
<pre>
Address Safer CPP warnings in ImageCapture.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=289257">https://bugs.webkit.org/show_bug.cgi?id=289257</a>
<a href="https://rdar.apple.com/146398834">rdar://146398834</a>

Reviewed by Chris Dumez.

* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::takePhoto):
(WebCore::ImageCapture::getPhotoCapabilities):
(WebCore::ImageCapture::getPhotoSettings):
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/292097@main">https://commits.webkit.org/292097@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0eda4d4d3260d9de141f88d0a6eb9562e9b2e35

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14414 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4261 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99839 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45312 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96871 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22835 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72328 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29626 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97825 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10948 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85618 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52660 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10646 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3337 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44652 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80870 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3432 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101882 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21856 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15958 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81327 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22103 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81639 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80709 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20189 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25276 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2677 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15084 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21831 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26948 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21491 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24962 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23231 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -79,16 +79,16 @@ void ImageCapture::takePhoto(PhotoSettings&& settings, DOMPromiseDeferred<IDLInt
-    m_track->takePhoto(WTFMove(settings))->whenSettled(RunLoop::protectedMain(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
-        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)] () mutable {
+    m_track->takePhoto(WTFMove(settings))->whenSettled(RunLoop::protectedMain(), [protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
+        queueTaskKeepingObjectAlive(protectedThis.get(), TaskSource::ImageCapture, [protectedThis = WTFMove(protectedThis), promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)] () mutable {
```
nit: likely doesn't need `.get()`

eric-carlson replied:
I get the following errors without it:

```
/Volumes/Work/WebKit/Sources/OpenSource/Source/WebCore/dom/ActiveDOMObject.h:111:49: error: 'WebCore::ActiveDOMObject::makePendingActivity' is not a member of class 'WTF::Ref<WebCore::ImageCapture>'
  111 |         auto activity = object.ActiveDOMObject::makePendingActivity(object);
      |                                ~~~~~~~~~~~~~~~~~^

/Volumes/Work/WebKit/Sources/OpenSource/Source/WebCore/Modules/mediastream/ImageCapture.cpp:83:9: note: in instantiation of function template specialization 'WebCore::ActiveDOMObject::queueTaskKeepingObjectAlive<WTF::Ref<WebCore::ImageCapture>>' requested here
   83 |         queueTaskKeepingObjectAlive(protectedThis, TaskSource::ImageCapture, [protectedThis = WTFMove(protectedThis), promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)] () mutable {
      |         ^

/Volumes/Work/WebKit/Sources/OpenSource/Source/WebCore/dom/ActiveDOMObject.h:112:16: error: no member named 'queueTaskInEventLoop' in 'WTF::Ref<WebCore::ImageCapture>'
  112 |         object.queueTaskInEventLoop(source, [protectedObject = Ref { object }, activity = WTFMove(activity), task = WTFMove(task)] () {
      |         ~~~~~~ ^
```

cdumez replied:
Oh, it's a templated function. Never mind.

jyavenard commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -108,19 +109,20 @@ void ImageCapture::getPhotoCapabilities(DOMPromiseDeferred<IDLDictionary<PhotoCa
-    m_track->getPhotoCapabilities()->whenSettled(RunLoop::protectedMain(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
-        legacyQueueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)] () mutable {
+    Ref protectedThis = Ref { *this };
+    m_track->getPhotoCapabilities()->whenSettled(RunLoop::protectedMain(), [protectedThis = WTFMove(protectedThis), promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
```
doing `protectedThis = Ref { *this }` looks simpler to me than creating the object first and moving it in the capture

cdumez replied:
I totally agree with Jean Yves here and the same issue applies elsewhere in the PR. We don't do local variable protectors outside lambda captures anymore.

cdumez commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -80,15 +80,15 @@ void ImageCapture::takePhoto(PhotoSettings&& settings, DOMPromiseDeferred<IDLInt
 
     m_track->takePhoto(WTFMove(settings))->whenSettled(RunLoop::protectedMain(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
-        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)](auto& capture) mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [protectedThis = WTFMove(protectedThis), promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)](auto& capture) mutable {
```
Capturing `protectedThis = WTFMove(protectedThis)` here is not needed. the `capture` parameter is essentially a protectedThis (`this` is kept alive already, per the function name).

cdumez replied:
I refactored `queueTaskKeepingObjectAlive()` to address ALL of the `this` capture static analysis recently and we conflicted a bit.

cdumez commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -109,7 +109,7 @@ void ImageCapture::getPhotoCapabilities(DOMPromiseDeferred<IDLDictionary<PhotoCa
 
     m_track->getPhotoCapabilities()->whenSettled(RunLoop::protectedMain(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
-        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)](auto& capture) mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [protectedThis = WTFMove(protectedThis), promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)](auto& capture) mutable {
```
ditto.

cdumez commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -138,7 +138,7 @@ void ImageCapture::getPhotoSettings(DOMPromiseDeferred<IDLDictionary<PhotoSettin
 
     m_track->getPhotoSettings()->whenSettled(RunLoop::protectedMain(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
-        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)](auto& capture) mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [protectedThis = WTFMove(protectedThis), promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)](auto& capture) mutable {
```
ditto.

eric-carlson committed:
Address Safer CPP warnings in ImageCapture.cpp
https://bugs.webkit.org/show_bug.cgi?id=289257
rdar://146398834

Reviewed by Chris Dumez.

* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::takePhoto):
(WebCore::ImageCapture::getPhotoCapabilities):
(WebCore::ImageCapture::getPhotoSettings):
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:

Canonical link: https://commits.webkit.org/292097@main
